lib/rouge/lexers/jinja.rb

```diff
@@ -36,6 +36,16 @@ def self.word_operators
       end
 
       state :root do
+        # Raw and verbatim blocks
+        open_raw_tags  = %r{(\{\%)(\s+)(verbatim|raw)(\s+)(\%\})}
+        close_raw_tags = %r{(\{\%)(\s+)(endverbatim|endraw)(\s+)(\%\})}
+
+        rule /#{open_raw_tags}(.*?)#{close_raw_tags}/m do
+          groups Comment::Preproc, Text, Keyword, Text, Comment::Preproc,
+                 Text,
+                 Comment::Preproc, Text, Keyword, Text, Comment::Preproc
+        end
+
```

Comment:
@robin850 Apologies for the length but here's an alternative to avoid having that regex at the start of `:root`:

```rb
# -*- coding: utf-8 -*- #
# frozen_string_literal: true

module Rouge
  module Lexers
    class Jinja < TemplateLexer
      title "Jinja"
      desc "Django/Jinja template engine (jinja.pocoo.org)"

      tag 'jinja'
      aliases 'django'

      mimetypes 'application/x-django-templating', 'application/x-jinja',
                'text/html+django', 'text/html+jinja'

      def self.keywords
        @@keywords ||= %w(as context do else extends from ignore missing
                          import include reversed recursive scoped
                          autoescape endautoescape block endblock call endcall
                          filter endfilter for endfor if endif macro endmacro
                          set endset trans endtrans with endwith without)
      end

      def self.tests
        @@tests ||= %w(callable defined divisibleby equalto escaped even iterable
                       lower mapping none number odd sameas sequence string
                       undefined upper)
      end

      def self.pseudo_keywords
        @@pseudo_keywords ||= %w(true false none True False None)
      end

      def self.raw_keywords
        @@raw_keywords ||= %w(raw verbatim)
      end

      def self.word_operators
        @@word_operators ||= %w(is in and or not)
      end

      state :root do
        # Comments
        rule /{#/, Comment, :comment

        # Statements
        rule /\{\%/ do
          token Comment::Preproc
          push :statement
        end

        # Expressions
        rule /\{\{/ do
          token Comment::Preproc
          push :expression
        end

        rule(/(.+?)(?=\\|{{|{%|{#)/m) { delegate parent }
        rule(/.+/m) { delegate parent }
      end

      state :filter do
        # Filters are called like variable|foo(arg1, ...)
        rule /(\|)(\w+)/ do
          groups Operator, Name::Function
        end
      end

      state :function do
        rule /(\w+)(\()/ do
          groups Name::Function, Punctuation
        end
      end

      state :text do
        rule /\s+/m, Text
      end

      state :literal do
        # Strings
        rule /"(\\.|.)*?"/, Str::Double
        rule /'(\\.|.)*?'/, Str::Single

        # Numbers
        rule /\d+(?=}\s)/, Num

        # Arithmetic operators (+, -, *, **, //, /)
        # TODO : implement modulo (%)
        rule /(\+|\-|\*|\/\/?|\*\*?)/, Operator

        # Comparisons operators (<=, <, >=, >, ==, ===, !=)
        rule /(<=?|>=?|===?|!=)/, Operator

        # Punctuation (the comma, [], ())
        rule /,/,  Punctuation
        rule /\[/, Punctuation
        rule /\]/, Punctuation
        rule /\(/, Punctuation
        rule /\)/, Punctuation
      end

      state :comment do
        rule(/[^{#]+/m) { token Comment }
        rule(/#}/) { token Comment; pop! }
      end

      state :expression do
        rule /\w+\.?/m, Name::Variable

        mixin :filter
        mixin :function
        mixin :literal
        mixin :text

        rule /%}|}}/, Comment::Preproc, :pop!
      end

      state :statement do
        rule /(\w+\.?)/ do |m|
          if self.class.keywords.include?(m[0])
            groups Keyword
          elsif self.class.pseudo_keywords.include?(m[0])
            groups Keyword::Pseudo
          elsif self.class.word_operators.include?(m[0])
            groups Operator::Word
          elsif self.class.tests.include?(m[0])
            groups Name::Builtin
          elsif self.class.raw_keywords.include?(m[0])
            token Keyword
            goto :raw
          else
            groups Name::Variable
          end
        end

        mixin :filter
        mixin :function
        mixin :literal
        mixin :text

        rule /\%\}/, Comment::Preproc, :pop!
      end

      state :raw do
        rule %r{(\{\%)(\s+)(endverbatim|endraw)(\s+)(\%\})} do
          groups Comment::Preproc, Text::Whitespace, Keyword, Text::Whitespace,
                 Comment::Preproc
          pop!
        end

        rule //, Text
      end
    end
  end
end
```

@dblessing Is this more what you had in mind?

Reply:
Hi @pyrmont,

Thanks for your proposal ; I actually had to tweak it a bit to make it work:

~~~diff
@@ -117,7 +117,10 @@

       state :statement do
         rule /(\w+\.?)/ do |m|
-          if self.class.keywords.include?(m[0])
+          if self.class.raw_keywords.include?(m[0])
+            token Keyword
+            goto :raw
+          elsif self.class.keywords.include?(m[0])
             groups Keyword
           elsif self.class.pseudo_keywords.include?(m[0])
             groups Keyword::Pseudo
@@ -125,9 +128,6 @@
             groups Operator::Word
           elsif self.class.tests.include?(m[0])
             groups Name::Builtin
-          elsif self.class.raw_keywords.include?(m[0])
-            token Keyword
-            goto :raw
           else
             groups Name::Variable
           end
@@ -142,13 +142,22 @@
       end

       state :raw do
+        mixin :text
+
+        rule /\%\}/ do
+          token Comment::Preproc
+          push :raw_block
+        end
+      end
+
+      state :raw_block do
         rule %r{(\{\%)(\s+)(endverbatim|endraw)(\s+)(\%\})} do
           groups Comment::Preproc, Text::Whitespace, Keyword, Text::Whitespace,
                  Comment::Preproc
-          pop!
+          pop!(2)
         end

-        rule //, Text
+        rule /(.+?)/, Text
       end
     end
   end
~~~

What do you think ? It's not ideal to have two states for that but since we are using `goto`, [this](https://github.com/rouge-ruby/rouge/blob/f494b53d9b04321032b9dc8a754e9f61198ec5f3/lib/rouge/lexers/jinja.rb#L134) is never triggered so the `%}` attached to the open tag doesn't get highlighted. Sorry, it's been a long time since I worked on that so this can certainly be improved.

(Off topic but since you three are already all notified though this issue, just wanted to say thank you for your work on this project folks ❤️ )